### ISSUE TITLE
IVY Error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class

### ISSUE THREAD

**Issue by stevethemacguy** [angular/angular#35399]
# 🐞 bug report

### Affected Package

 The issue might be related to package 
[ng-packagr](https://github.com/ng-packagr/ng-packagr)

### Is this a regression?

 Yes, the previous version in which this bug was not present was: 8.2.14

### Description
Possible duplicate of #35331, #34983, #35255, and #30972, but I'm including a minimal reproduction repo. Building with IVY throws the following error (see the Exception or Error section)

![StackBlitz Error](https://user-images.githubusercontent.com/1767135/74392714-e0448a00-4dbc-11ea-8bd8-a938995b0394.png)

Our real app, which uses a shared module, throws a slightly different error:

![Concise Error](https://user-images.githubusercontent.com/1767135/74392793-1a159080-4dbd-11ea-9cc8-ba873769fefd.png)

Running the ng-packagr postinstall script adds "__ivy_ngcc__" folders to some of our other dependencies, but does not touch the [angular2-highcharts](https://github.com/gevgeny/angular2-highcharts) library. The library does not use the Angular Package Format, which might be part of the problem, but I'm not the library's author and have never built an ES6/angular library myself, so I'm just guessing at a potential fix.

Unfortunately, the library is no longer maintained, but it is one of our app's core/major dependencies. I also attempted to rebuild the library from the source files after adding the following to the library's package.json file:

<pre><code>
"postinstall": "ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points",
"ngPackage": {
  "lib": {
    "entryFile": "index.d.ts"
  }
},
</pre></code>

But this produces the following error:

![Error](https://user-images.githubusercontent.com/1767135/74393168-20583c80-4dbe-11ea-8024-d00a1ff2c4cc.png)

There is no "files" array provided in the library's package.json file, so I'm not sure what the error is referring to. I fixed all Angular 9-related issues prior to re-building the library, so I assume it has to do with the way it's packaged or how it exports the modules/components.

## 🔬 Minimal Reproduction

StackBlitz
https://angular-ivy-cant-be-resolved.stackblitz.io

GitHub repo in case StackBlitz isn't working:
https://github.com/stevethemacguy/angular-ivy-cant-be-resolved-bug

Steps to Reproduce
1. Download or clone the repo
2. npm install
3. ng build (or ng serve)

Expected behavior: IVY builds the app
Actual behavior: IVY displays an error.

## 🔥 Exception or Error
<pre><code>ERROR in node_modules/angular2-highcharts/dist/index.d.ts:8:22 - error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class

8 export declare class ChartModule {
</code></pre>

## 🌍  Your Environment
**Angular Version:**
<pre><code>Angular CLI: 9.0.2
Node: 10.15.3
OS: darwin x64

Angular: 9.0.1
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.2
@angular-devkit/build-angular     0.900.2
@angular-devkit/build-optimizer   0.900.2
@angular-devkit/build-webpack     0.900.2
@angular-devkit/core              9.0.2
@angular-devkit/schematics        9.0.2
@angular/cli                      9.0.2
@ngtools/webpack                  9.0.2
@schematics/angular               9.0.2
@schematics/update                0.900.2
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2

</code></pre>

**Anything else relevant?**
After disabling IVY, our real app successfully builds in AOT (prod mode) using Angular 9.0.1, so the issue seems to be an Angular regression that is un-related to our app (except for the [angular2-highcharts](https://github.com/gevgeny/angular2-highcharts) dependency). Our app uses a custom angular/webpack build. It does not use Angular CLI.

**Comment by robertlimanto** [comment]
any idea on this error?

**Comment by houdass** [comment]
Hi guys, I figured out what going wrong, I had a Angular 8 project (containing a library project) that I have upgraded to Angular 9 (with a lib also: ng generate library core, I created at first a new angular 9 project to compare the differences with my project made with ng 8, i figure out that there is some things missing, here is the steps I followed:
- Delete node_modules
- Delete dist folder
- In the Angular 9 project lib there is a new file called: tsconfig.lib.prod.json 
```
{
  "extends": "./tsconfig.lib.json",
  "angularCompilerOptions": {
    "enableIvy": false
  }
}
```
- In your angular.json add the following:
```
{
  ...
  "projects": {
     ...
     "configurations": {
            "production": {
              "tsConfig": "projects/core/tsconfig.lib.prod.json"
             }
      }
  ...
  }
}
```
- run ng build --prod, and everything will work :D

PS: in this case the library is called "core"

**Comment by gmanojisaac** [comment]
@houdass
Issue is still there..
I have done the following: 
npm uninstall -g @angular/cli
npm cache verify
npm install -g @angular/cli@latest
ng --version => **Angular CLI: 9.0.2**
ng new angularLatest
cd angularLatest
Inside app.module => import { MatButton } from '@angular/material/button';
added in the list of imports => MatButton 
ng build
**Error** : ERROR in node_modules/@angular/material/button/button.d.ts:22:22 - error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class

**Comment by gmanojisaac** [comment]
@houdass
Issue is still there..
I have done the following:
npm uninstall -g @angular/cli
npm cache verify
npm install -g @angular/cli@latest
ng --version => Angular CLI: 9.0.2
ng new angularLatest
cd angularLatest
ng add @angular/material
Inside app.module => import { MatButton } from '@angular/material/button';
added in the list of imports => MatButton
ng build
Error : ERROR in node_modules/@angular/material/button/button.d.ts:22:22 - error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class

**Comment by alienjedi** [comment]
@houdass I want to try your fix. But, I cannot see any tsconfig.lib.prod.json file. And I am not building a library. My NG6002 error is coming from primeNG library. A component library that I am using to build my app. I am lost as to the fix you are proposing. Please explain further.

**Comment by gkalpak** [comment]
Thx for sharing your solution, @houdass. However, yours seems to be different issue that than the one reported here.

**Comment by AliAdravi** [comment]
Rebuild the application again and started working!

**Comment by ImmutableVector** [comment]
Not sure this will provide anyone relief or assist with figuring out what the root cause is, but cleaning my cache seemed to hlep resolve the issue.
 `npm cache clean --force`

 The only other step that I did was I ran a --prod serve after the clean.
`ng serve --prod`
 
I do not know if the clean or --prod thing had anything to do with fixing it for me, or if maybe the combination of them did anything, but I can no longer repro this bug in my project.

Other than that, I saw this first when adding HttpClientModule as I was going through the tour of heroes tutorial.

**Comment by flash-me** [comment]
Hi @stevethemacguy ,

I just encountered the same issue and was able to track it down to **one property** in the **tsconfig.json**:
`Ensure that the tsconfig.json of your library has set "importHelpers": true`

What I tried with **importHelper** in their corresponding **tsconfig.json**:
| Angular Library | Consuming App | Result |
| --- | --- | --- |
| true | true | works! |
| true | false | works! |
| false | true | fails! |
| false | false | fails! |

Cheers
flash ⚡️

**Comment by FranzZemen** [comment]
When get NG6002 with a module issue as above, delete the top node_modules in the workspace (where you should have installed it), and rebuild libraries/apps.

Seems like on first build, output is added to node_modules that gets occasionally confused.

This 'trick" works for me with problems incrementally added angular material modules to a library and app.

**Comment by flash-me** [comment]
Ok, more details on this!

- So setting **"importHelpers": true** solves the problem **always**. 
- When setting to **false** not only your library will break. Also I observed that the bundle size will **explore**

I thought setting importHelpers to false would cause the typescript compiler to include the helper methods into the bundle. And indeed this is happing. **But not only once!!**


A quick look into the created bundles reveals following with **importHelpers set to false**:

### UMD Bundle
![image](https://user-images.githubusercontent.com/5240450/74611141-8f54ce80-50f9-11ea-9466-25a792a78985.png)


### FESM2015 Bundle
![image](https://user-images.githubusercontent.com/5240450/74611117-7ba96800-50f9-11ea-90f9-c4fe2fb6aaf3.png)


Maybe this is also the reason why the bundle sizes grew?

**Comment by paksouse** [comment]
Hello, 
I have the same problem in my poject "NX monorepo project", 

- "importHelpers": true didnot help

-   "enableIvy": false for the libs and  "enableIvy": true  for the demo project that inject the libs dinot work either.

Any suggestions please ^^
regards.

**Comment by FranzZemen** [comment]
@paksouse, see my comment above in the thread.  It has helped me and
others.  Unfortunately it looks like a couple of conditions can drive
NG6002...if you have an issue when adding modules over time with Angular 9
then my workaround should do it.

**Comment by rafastellato** [comment]
Hello,

Same problem here. I tried to delete node_modules and reinstall packages again but didn't work. The "importHelpers" for me was true already.

I need start a new angular project and I am thinking if I should start with Angular 9

**Comment by paksouse** [comment]
@FranzZemen thanks for your suggestion but didnot works.
ERROR in Failed to compile entry-point @project/lib (es2015 as esm2015) due to compilation errors:
libs/.../...module.ts:7:14 - error NG6002: Appears in the NgModule.imports of testModule, but could not be resolved to an NgModule class

**Comment by rafastellato** [comment]
For me, disable Ivy works! (To disable Ivy, you need add a flag "enableIvy" false in the file tsconfig.app.json. More information here: https://angular.io/guide/ivy)

But, I dont know the consequences of doing this, and when I would can enable again.

**Comment by flash-me** [comment]
@paksouse and you are sure that you enabled importHelpers for your library? That's strange because at work I tried even more library repositories and this definitely solved the issues.. Hm.

and as @rafastellato mentioned, for libraries ivy must be disabled, which is already told in the docs, though.
And with 9.x.x ivy is enabled by default, which means existing libs must set it to false explicitly.

Another point: For libraries installed by npm, you can't really set importHelpers to false, so you must struggle around with deleting node_modules, if it helps.

**Comment by Embraser01** [comment]
For those who `"importHelpers": true` didn't solve the problem and are using yarn workspaces:

You need to specify the node_module folder in the `postinstall` script.

Ex: If you have a workspace with the angular project in `packages/client/package.json` you have to set the postinstall script to:
 `ngcc --source ../../node_modules --properties es2015 browser module main --first-only --create-ivy-entry-points` 

For more options you can go [here].(https://github.com/angular/angular/blob/master/packages/compiler-cli/ngcc/main-ngcc.ts#L20)

**Comment by stevethemacguy** [comment]
Thank you for the responses, but I am not using Angular CLI and adding "importHelpers" and deleting node modules has no effect.

**Comment by paksouse** [comment]
thank you for your answers @Embraser01  @flash-me 
Its works for me localy even with a huge warning about "typescript.js 5203:41-60
Critical dependency: the request of a dependency is an expression" but i get the same error "NG6002" in the jenkins server!

**Comment by gkalpak** [comment]
@stevethemacguy, it turns out that the problem is three-fold:

1. The Ivy compatibility compiler (ngcc) did not correctly recognize a particular pattern used in `angular2-highcharts` (see below).
2. The `angular2-highcharts` package does not exactly follow the recommended package format for Angular libraries, which prevents ngcc from realizing it needs to process the package.
3. The demo app is using private exports of `angular2-highcharts`, which it shouldn't.

The first two will be fixed by #35527. You can try out the fixes by grabbing the build artifacts from [here][1] (see [there][2] for instructions on how to use them).

For the third issue, I changed [app/app.module.ts][3] as shown below and the app seems to work fine (with the changes from #35527):

```diff
 import { NgModule } from '@angular/core';
 import { BrowserModule } from '@angular/platform-browser';
 import { FormsModule } from '@angular/forms';
 import { ChartModule } from 'angular2-highcharts';
 import { AppComponent } from './app.component';
-
-import { HighchartsStatic } from 'angular2-highcharts/dist/HighchartsService';
 
 
 // This just includes an extra HighCharts "exporting" module.
 export function highchartsFactory() {
   const hc = require('highcharts');
   const exporting = require('highcharts/modules/exporting');
   exporting(hc);
   return hc;
 }
 
 @NgModule({
   imports: [
     BrowserModule,
     FormsModule,
     // Import ChartModule from the angular2-highcharts library
-    ChartModule,
+    ChartModule.forRoot(highchartsFactory()),
   ],
-  providers: [
-    {
-      // Provide the HighCarts service
-      provide: HighchartsStatic,
-      useFactory: highchartsFactory
-    },
-  ],
   declarations: [AppComponent],
   bootstrap: [AppComponent]
 })
 export class AppModule { }
```

##
@stevethemacguy, it would be great if you could try the fixes with your actual app and report back whether the problem is actually fixed or not.

[1]: https://circleci.com/gh/angular/angular/628101#artifacts/containers/0
[2]: https://github.com/angular/angular/blob/183a8629a/docs/DEVELOPER.md#getting-packages-from-build-artifacts
[3]: https://github.com/stevethemacguy/angular-ivy-cant-be-resolved-bug/blob/29da24065/src/app/app.module.ts

**Comment by pablo-the-souza** [comment]
I had the same issue with firestore and the only solution was to opt out of Ivy. I tried reinstalling node_modules again and it didn't work. 

[https://stackoverflow.com/questions/60290309/error-ng6002-appears-in-the-ngmodule-imports-of-appmodule-but-could-not-be-res](url)

**Comment by paksouse** [comment]
Hey, Finally i decided to disable Ivy to build in '--prod'.
I have my last question : Is it advisable to enable Ivy for ng build --prod or not ?
regards.

**Comment by flash-me** [comment]
> Hey, Finally i decided to disable Ivy to build in '--prod'.
> I have my last question : Is it advisable to enable Ivy for ng build --prod or not ?
> regards.

- For **angular libraries** the recommendation is to **disable** Ivy when publishing in order To stay compatible as long as possible. 
- For **angular apps** you may **enable** Ivy, so you can benefit from the advantages of the Ivy Compiler. Every library used in the app will also be compiled, if not already.

**Comment by paksouse** [comment]
@flash-me okeyy thank you^^

**Comment by chaosmonster** [comment]
Similar issue @flash-me suggestion to use `importHelpers` solved it.
My reproduction case:

```
ng new test; cd test/;
ng add @angular/material;
```

then add `MatSliderModule` to `app.module.ts` and after that add `MatButtonModule` to `app.module.ts`.

I get the said error `NG6002`

**Comment by gkalpak** [comment]
BTW, this should be fixed by #35191 (although using `importHelpers: true` is the recommended way anyway).

**Comment by stevethemacguy** [comment]
@gkalpak, Thank you for the prompt response! So many people were posting on this issue that I actually missed your response. I'll try out the changes and report back, but I probably won't get to it before Monday. Thanks!

**Comment by raulicious** [comment]
Approximately when will this fix be merged and available?

**Comment by gkalpak** [comment]
The fix has been merged and will be included in the next release (9.0.3).
<sub>Patch releases usually happen weekly (usually on Wednesdays/Thursdays).</sub>

**Comment by chulphan** [comment]
> Not sure this will provide anyone relief or assist with figuring out what the root cause is, but cleaning my cache seemed to help resolve the issue.
> `npm cache clean --force`
> 
> The only other step that I did was I ran a --prod serve after the clean.
> `ng serve --prod`
> 
> I do not know if the clean or --prod thing had anything to do with fixing it for me, or if maybe the combination of them did anything, but I can no longer repro this bug in my project. Note, I can run without --prod and still have successful builds going forward.
> 
> Other than that, I saw this first when adding HttpClientModule as I was going through the tour of heroes tutorial.

It works for me.

**Comment by candyman00** [comment]
I'm still facing this issue.

I'm currently developing a library with ivy disabled ( as suggested ).

When i use it on an app, the error appears: 

```
2801 export declare class HttpClientModule {
                          ~~~~~~~~~~~~~~~~
../components/node_modules/@angular/common/http/http.d.ts:2801:22 - error NG6002: Appears in the NgModule.imports of SBCoAuthModule, but could not be resolved to an NgModule class.

This likely means that the library (@angular/common/http) which declares HttpClientModule haes HttpClientModule has not been processed correctly by ngcc, or is noof the library is avait compatible with Angular Ivy. Check if a newer version of the library is available, and update if so. Also considerer version of the library is available, and uonsider checking wipected to be compatible with Ivy.pdate if so. Also consider checking with the library's authors to see if the library is expected to be compatible with Ivy.

```

Both, the app and the library are updated to angular 9.0.3, and i use the latest version of ng-packagr as well.

i've tried to use ivy on the library as well, but isnt work either.

Also, i've tried to run the ngcc as a postintall script on the app and as well after the library build, over the dist folder. But the error remains.

This works fine with ivy disabled on both, app and library.

any help?

thanks in advance!! :)

**Comment by stevethemacguy** [comment]
@gkalpak I updated to 9.0.4 and confirmed that the angular2-highcharts package now builds (and runs) successfully with IVY. Thank you! 🙇

**Comment by andreialecu** [comment]
@candyman00 Do you by any chance use `yarn workspaces` in the app? I'm also troubleshooting the exact same error in an angular library here: https://github.com/aitboudad/ngx-loading-bar/issues/151

I cannot seem to reproduce it in a fresh new angular project though - seems to compile properly there. In the main app we use yarn workspaces though.

**Comment by candyman00** [comment]
@andreialecu  nope, but i'm using a symlink between the library and the app ( wich works fine disabling ivy in both app and demo )

i've just update to angular 9.0.4, an the error remains. So i dont know what to do now :S

**Comment by andreialecu** [comment]
@candyman00 then it's most likely the same as https://github.com/angular/angular/issues/35447 but also related #35747 as fixing that would probably fix both.

**Comment by pleinonen** [comment]
@andreialecu I am also using yarn workspaces and this [one](https://github.com/angular/angular/issues/35399#issuecomment-587181811) fixed issue for me. I did add the postinstall script to client's package.json (projects/client/package.json) and cleared all node_module folders before reinstall. It should be also noted that in my project all Angular dependencies are located in root package.json as I was not able to make CLI work otherwise.

**Comment by alePanta** [comment]
In my case was a library (PrimeNG) the problem. I updated it to the last version and all works fine.

**Comment by daminiCES** [comment]
Facing the same problem with Angular 9.0.4. Here is the scenario:
Working on an Angular 9.0.4 project which contains a library (my-lib) and an application. The library is being used in the same project as a npm dependency.
Here are the steps:
1) Created a library (by exporting the module, component & services, "importHelpers": true)
2) Build the library with --prod `"enableIvy": false`
3) Packed the library:
```
cd dist/my-lib
npm pack
```
4) Installed the library in the project:
`npm install dist/my-lib/my-lib-0.0.1.tgz`
5) Imported the Module form the library:
`import { MyLibModule } from 'my-lib';`
6) Build the project --prod `"enableIvy": true`
Getting the following error:
![image](https://user-images.githubusercontent.com/13457141/75771806-b1dd1d80-5d70-11ea-9197-7ef81f784de9.png)

**Any suggestions?**
_Note: The build works fine if we build the project with ` "enableIvy": false`_

**Comment by flash-me** [comment]
@daminiCES hi,

did you set a path mapping in your app tsconfig?

I realized that ngcc does not process relativ paths

cheers
flash :zap:

**Comment by daminiCES** [comment]
@flash-me Hi,
I didn't set the path, however, the default path was set to this:
```
"paths": {
      "my-lib": [
        "dist/my-lib/my-lib",
        "dist/my-lib"
      ]
    }
```
Which was not correct in my case(because we are installing the library as a package). After removing it, the build passed!
Thanks for your help! 👍

**Comment by mojo722** [comment]
> @daminiCES hi,
> 
> did you set a path mapping in your app tsconfig?
> 
> I realized that ngcc does not process relativ paths
> 
> cheers
> flash ⚡️

Are you saying that if I am using PATHS in the tsconfig and in my components then I will get this issue when importing other modules?

**Comment by flash-me** [comment]
> > @daminiCES hi,
> > did you set a path mapping in your app tsconfig?
> > I realized that ngcc does not process relativ paths
> > cheers
> > flash zap
> 
> Are you saying that if I am using PATHS in the tsconfig and in my components then I will get this issue when importing other modules?

**relative** PATHS, yes, at least for me

`'../foo-lib/foo.module.ts'` errors, because ngcc cannot resolve it

but setting a PATH mapping in the tsconfig of the consuming app like 
```JSON
// ...
paths: {
  "foo-lib": [
    "../foo-lib"
  ],
  "foo-lib/*": [
    "../foo-lib/*"
  ]
}
// ...
```

and in the consuming app using `import { FooModule } from 'foo-lib';`
worked for me. (And actually for others, too).

cheers
flash :zap:

**Comment by nachbar** [comment]
> Not sure this will provide anyone relief or assist with figuring out what the root cause is, but cleaning my cache seemed to help resolve the issue.
> `npm cache clean --force`

That was my experience as well.  I had added FlexLayoutModule to an angular/material project with FlexLayout directives, and got this error until I used `npm cache clean --force`

**Comment by dyingangel666** [comment]
Guys you should set` importHelpers: true` in the tsconfig of your library. And don't forget to set `enableIvy:false` as well. Follow the update docs and changelog of angular and update the stuff manually to make sure all stuff is correct. I've also seen that ng update isn't updating all stuff correctly. I've just setted up a new library project with the CLI and than compared the tsconfigs etc.

**Comment by flash-me** [comment]
> Guys you should set` importHelpers: true` in the tsconfig of your library. And don't forget to set `enableIvy:false` as well. Follow the update docs and changelog of angular and update the stuff manually to make sure all stuff is correct. I've also seen that ng update isn't updating all stuff correctly. I've just setted up a new library project with the CLI and than compared the tsconfigs etc.
> 
> Then all is running like a charm ;-) Here's my tsconfig that is running without any problems:
> 
> ```json
> ...
>  "angularCompilerOptions": {
>     "enableIvy": false,
>     "skipTemplateCodegen": true,
>     "strictMetadataEmit": true,
>     "enableResourceInlining": true
>   },
> "compilerOptions": {
>     "sourceMap": true,
>     "declaration": false,
>     "downlevelIteration": true,
>     "experimentalDecorators": true,
>     "module": "esnext",
>     "moduleResolution": "node",
>     "importHelpers": true,
>     "target": "es2015",
>     "typeRoots": [
>       "node_modules/@types"
>     ],
>     "lib": [
>       "es2018",
>       "dom"
>     ]
>   }
> ...
> ```

https://github.com/angular/angular/issues/35399#issuecomment-586742346

Cheers
flash :zap:

**Comment by Mautriz** [comment]
Gives me this error every time I add a new material module and I have to run the cache clean (even for dev mode), angular 9.0.5, components 9.1.1

**Comment by dhubens** [comment]
> I'm still facing this issue.
> 
> I'm currently developing a library with ivy disabled ( as suggested ).
> 
> When i use it on an app, the error appears:
> 
> ```
> 2801 export declare class HttpClientModule {
>                           ~~~~~~~~~~~~~~~~
> ../components/node_modules/@angular/common/http/http.d.ts:2801:22 - error NG6002: Appears in the NgModule.imports of SBCoAuthModule, but could not be resolved to an NgModule class.
> 
> This likely means that the library (@angular/common/http) which declares HttpClientModule haes HttpClientModule has not been processed correctly by ngcc, or is noof the library is avait compatible with Angular Ivy. Check if a newer version of the library is available, and update if so. Also considerer version of the library is available, and uonsider checking wipected to be compatible with Ivy.pdate if so. Also consider checking with the library's authors to see if the library is expected to be compatible with Ivy.
> ```
> 
> Both, the app and the library are updated to angular 9.0.3, and i use the latest version of ng-packagr as well.
> 
> i've tried to use ivy on the library as well, but isnt work either.
> 
> Also, i've tried to run the ngcc as a postintall script on the app and as well after the library build, over the dist folder. But the error remains.
> 
> This works fine with ivy disabled on both, app and library.
> 
> any help?
> 
> thanks in advance!! :)

Hey @candyman00 , could you resolve your problem? I have the exact same one and none of the workarounds in this thread did help me :(

**Comment by dhubens** [comment]
Update: It works, when I remove the HttpClientModule from `imports: []` of my library and just add it to my Project's `imports`. It seems absolutely wrong and just like a workaround or hack, but it works for now.

Just fyi: I have the problems only in conjunction with yarn link or npm link. Oh, and you have to use `ngcc` after linking.

**Comment by petebacondarwin** [comment]
> Just fyi: I have the problems only in conjunction with yarn link or npm link. Oh, and you have to use ngcc after linking.

@dhubens - I am afraid that ngcc does not process packages that have been symlinked into the `node_modules` directory. This is probably the cause of your errors. Instead generate a packed version of your library and then install that into your application project.

**Comment by candyman00** [comment]
@dhubens Well yes but no haha

I forgot to mentioned that i've already set the `importHelpers: true`.

I maked my library works, but because we are not working in a monorepo project, so if you are in this case. this is not work for you at all :(

As many people said, IVY seems to not process symlinked deps. so i maked a trick to bypass this.

- remove node_modules from the app
- remove the library dependencie declaration from the package.json of the app
- npm i on the app
- compile the library ( ivy disabled )
- copy the content of the dist folder of the library to the node_modules folder in the app
- run the postinstall script manually ("postinstall": "ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points")


This creates the `__ivy_ngcc__` folder and everthing works fine (well some paths to the assets folder i break due the added level, but this is not releveant)


yeah, y knowk that this is not a solution at all, but i only need to see if the error is gone, because my library includes a demo app on where everithing works fine since the beginning, and i only need to see if the library works as a npm package.

Once i sed that the error is gone, i had proceed to publish the library on a private nexus and this works just fine. All the problem cames from the symlinked deps

**Comment by dhubens** [comment]
> > Just fyi: I have the problems only in conjunction with yarn link or npm link. Oh, and you have to use ngcc after linking.
> 
> @dhubens - I am afraid that ngcc does not process packages that have been symlinked into the `node_modules` directory. This is probably the cause of your errors. Instead generate a packed version of your library and then install that into your application project.

Thanks for your answer. Yes, I could package the library and publish it on npm, but I want to actively develop the library and can't publish a new version every time I want to test a new line of code. I mean there are so many libraries, how do they solve this?

**Comment by dhubens** [comment]
> @dhubens Well yes but no haha
> 
> I forgot to mentioned that i've already set the `importHelpers: true`.
> 
> I maked my library works, but because we are not working in a monorepo project, so if you are in this case. this is not work for you at all :(
> 
> As many people said, IVY seems to not process symlinked deps. so i maked a trick to bypass this.
> 
> * remove node_modules from the app
> * remove the library dependencie declaration from the package.json of the app
> * npm i on the app
> * compile the library ( ivy disabled )
> * copy the content of the dist folder of the library to the node_modules folder in the app
> * run the postinstall script manually ("postinstall": "ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points")
> 
> This creates the `__ivy_ngcc__` folder and everthing works fine (well some paths to the assets folder i break due the added level, but this is not releveant)
> 
> yeah, y knowk that this is not a solution at all, but i only need to see if the error is gone, because my library includes a demo app on where everithing works fine since the beginning, and i only need to see if the library works as a npm package.
> 
> Once i sed that the error is gone, i had proceed to publish the library on a private nexus and this works just fine. All the problem cames from the symlinked deps

Even if this works, it's just a dirty hack while developing. For every new line of code I would have to copy the dist folder manually and run ngcc. I mean ngcc does something when I use `preserveSymLinks: true` in the main project. But it's still not working when I import `HttpClientModule` within the library.

**Comment by dyingangel666** [comment]
@dhubens To simplify your process you can also use a npm tool to copy the files automatically after your ngcc command. Finally just wrap both command in a npm script in your package.json

**Comment by petebacondarwin** [comment]
> Yes, I could package the library and publish it on npm, but I want to actively develop the library and can't publish a new version every time I want to test a new line of code.

@dhubens - just to clarify... I am not proposing to publish the library. Just package it into a `tgz` archive file (via the command `npm pack` run inside the library folder). This can then be used one the same machine via `file://...` dependency in your test application package.json. For example:

```json
"dependencies": {
  ...
  "my-library": "file:///path/to/my-library/project/dist/my-library.tgz",
  ...
```

The main downside of this is that you need to rebuild and re-pack your library, and then re-install it into your test library each time you want to test a change. But generally, in my experience, you will work mostly with unit tests locally to the library for the most part, and then treat tests in your test project as an e2e test.

If you want to use the test project as more of a unit test environment, then the approach is to include the actual source of the library (TS files) directly via pathMappings in your test project. This avoids ngcc altogether as ngtsc will build the library as part of your application, allowing fast turnaround. The downside of this is that you are testing a direct ivy build of you library and not the ViewEngine build that will actually be deployed.

**Comment by SumailaBello** [comment]
> > Not sure this will provide anyone relief or assist with figuring out what the root cause is, but cleaning my cache seemed to help resolve the issue.
> > `npm cache clean --force`
> > The only other step that I did was I ran a --prod serve after the clean.
> > `ng serve --prod`
> > I do not know if the clean or --prod thing had anything to do with fixing it for me, or if maybe the combination of them did anything, but I can no longer repro this bug in my project. Note, I can run without --prod and still have successful builds going forward.
> > Other than that, I saw this first when adding HttpClientModule as I was going through the tour of heroes tutorial.
> 
> It works for me.

This worked for me

**Comment by clement911** [comment]
`npm cache clean --force` didn't help for us.

Deleting `node_modules` and running `npm install` also still results in the same error.

**Comment by SumailaBello** [comment]
> `npm cache clean --force` didn't help for us.
> 
> Deleting `node_modules` and running `npm install` also still results in the same error.

Try disabling ivy.

**Comment by clement911** [comment]
Disabling Ivy does work but then what's the point of upgrading to angular 9?
From what I understand, Ivy is THE big update of version 9...

**Comment by flash-me** [comment]
@clement911 https://github.com/angular/angular/issues/35399#issuecomment-588206596

 cheers
flash :zap:

**Comment by clement911** [comment]
As it turns out, for us the issue was caused by the presence of `node_modules` folder inside the `src` folder. Removing this folder made this error go away.
Hope it helps...

**Comment by flash-me** [comment]
> > Yes, I could package the library and publish it on npm, but I want to actively develop the library and can't publish a new version every time I want to test a new line of code.
> 
> @dhubens - just to clarify... I am not proposing to publish the library. Just package it into a `tgz` archive file (via the command `npm pack` run inside the library folder). This can then be used one the same machine via `file://...` dependency in your test application package.json. For example:
> 
> ```
> "dependencies": {
>   ...
>   "my-library": "file:///path/to/my-library/project/dist/my-library.tgz",
>   ...
> }
> ```
> 
> The main downside of this is that you need to rebuild and re-pack your library, and then re-install it into your test library each time you want to test a change. But generally, in my experience, you will work mostly with unit tests locally to the library for the most part, and then treat tests in your test project as an e2e test.
> 
> If you want to use the test project as more of a unit test environment, then the approach is to include the actual source of the library (TS files) directly via pathMappings in your test project. This avoids ngcc altogether as ngtsc will build the library as part of your application, allowing fast turnaround. The downside of this is that you are testing a direct ivy build of you library and not the ViewEngine build that will actually be deployed.
 
@petebacondarwin  

1. cd into dist folder of <lib>
2. `npm link`
3. go back In your app root
4. `npm link <lib>`
5. ???
6. profit 

Don't struggle around with packaging for developing locally. .. 

https://docs.npmjs.com/cli/link.html

cheers
flash :zap:

**Comment by neharica** [comment]
> > Yes, I could package the library and publish it on npm, but I want to actively develop the library and can't publish a new version every time I want to test a new line of code.
> 
> @dhubens - just to clarify... I am not proposing to publish the library. Just package it into a `tgz` archive file (via the command `npm pack` run inside the library folder). This can then be used one the same machine via `file://...` dependency in your test application package.json. For example:
> 
> ```
> "dependencies": {
>   ...
>   "my-library": "file:///path/to/my-library/project/dist/my-library.tgz",
>   ...
> }
> ```
> 
> The main downside of this is that you need to rebuild and re-pack your library, and then re-install it into your test library each time you want to test a change. But generally, in my experience, you will work mostly with unit tests locally to the library for the most part, and then treat tests in your test project as an e2e test.
> 
> If you want to use the test project as more of a unit test environment, then the approach is to include the actual source of the library (TS files) directly via pathMappings in your test project. This avoids ngcc altogether as ngtsc will build the library as part of your application, allowing fast turnaround. The downside of this is that you are testing a direct ivy build of you library and not the ViewEngine build that will actually be deployed.

@petebacondarwin We are trying to upgrade **monorepo to Angular V9** and are maintaining it by rush.js (https://rushjs.io/pages/intro/welcome/) and we are using **pnpm as a package manager**. I am facing the above mentioned IVY **Error NG6002 error while running tests of libraries**. (Have considered all the above suggestions but nothing helped) . From the documentation and the error that is shown during rush test, it is clear that the corresponding module(here in our case @angular/flex-layout) has not been processed correctly by the ngcc compiler. Documentation says we need to run "ngcc" command for the modules to be IVY compliant. But somehow pnpm is not picking up postinstall script(which includes ngcc command) present in package.json of each library and is not executed. Even though we are able to install and build(done by ng-packagr) the entire project which includes both libraries and demo app, we are blocked because of tests throwing this issue. Any suggestions or help around this would be greatly appreciated.

**Note:** Everything works fine individually meaning running install, build and test using yarn in each library and application works but collectively when we run rush test, we are facing this issue. (**Reason** : because we don't face symlinks issue when we run projects  individually)

**Comment by yamaha252** [comment]
app <- library1 <- library2 – doesn't work for me (library2 errors), although app which imports a library works fine

**Comment by petebacondarwin** [comment]
@yamaha252 - is this when ngcc processes the libraries? Do you use pathMappings to relate library2 to library1? If so, then may be this PR that just landed will help: https://github.com/angular/angular/pull/36180. But without more information about your specific scenario it is difficult to know.

**Comment by yamaha252** [comment]
@petebacondarwin nope, I just use `yarn link library1` in my app and `library1` is built with `library2` which is installed from repo. `library2` is also installed from repo in my app.
If I use installed `library1` (not a link) everything is allright.

Errors are on compilation `library1` (es2015 as esm2015) 

```
ERROR in Failed to compile entry-point @cbdev/ui-kit (es2015 as esm2015) due to compilation errors:
../angular-mfa/node_modules/@angular/cdk/a11y/a11y-module.d.ts:9:22 - error NG6002: Appears in the NgModule.imports of FocusModule, but could not be resolved to an NgModule class.

This likely means that the library (@angular/cdk/a11y) which declares A11yModule has not been processed correctly by ngcc, or is not compatible with Angular Ivy. Check if a newer version of the library is available, and update if so. Also consider checking with the library's authors to see if the library is expected to be compatible with Ivy.

9 export declare class A11yModule {
                       ~~~~~~~~~~
../angular-mfa/node_modules/@angular/common/http/http.d.ts:2801:22 - error NG6002: Appears in the NgModule.imports of IconModule, but could not be resolved to an NgModule class.

This likely means that the library (@angular/common/http) which declares HttpClientModule has not been processed correctly by ngcc, or is not compatible with Angular Ivy. Check if a newer version of the library is available, and update if so. Also consider checking with the library's authors to see if the library is expected to be compatible with Ivy.

2801 export declare class HttpClientModule {
                          ~~~~~~~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:153:5 - error NG6002: Appears in the NgModule.imports of ButtonIconModule, but itself has errors

153 let IconModule = class IconModule {
        ~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:153:5 - error NG6002: Appears in the NgModule.imports of InputModule, but itself has errors

153 let IconModule = class IconModule {
        ~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:169:5 - error NG6002: Appears in the NgModule.imports of InputModule, but itself has errors

169 let ButtonIconModule = class ButtonIconModule {
        ~~~~~~~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:42:5 - error NG6002: Appears in the NgModule.imports of InputModule, but itself has errors

42 let FocusModule = class FocusModule {
       ~~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:42:5 - error NG6002: Appears in the NgModule.imports of InputModule, but itself has errors

42 let FocusModule = class FocusModule {
       ~~~~~~~~~~~
../angular-mfa/node_modules/@angular/cdk/a11y/a11y-module.d.ts:9:22 - error NG6002: Appears in the NgModule.imports of SelectModule, but could not be resolved to an NgModule class.

This likely means that the library (@angular/cdk/a11y) which declares A11yModule has not been processed correctly by ngcc, or is not compatible with Angular Ivy. Check if a newer version of the library is available, and update if so. Also consider checking with the library's authors to see if the library is expected to be compatible with Ivy.

9 export declare class A11yModule {
                       ~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:169:5 - error NG6003: Appears in the NgModule.exports of FormModule, but itself has errors

169 let ButtonIconModule = class ButtonIconModule {
        ~~~~~~~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:670:5 - error NG6003: Appears in the NgModule.exports of FormModule, but itself has errors

670 let InputModule = class InputModule {
        ~~~~~~~~~~~
../angular-mfa/node_modules/@cbdev/ui-kit/fesm2015/ui-kit.js:1186:5 - error NG6003: Appears in the NgModule.exports of FormModule, but itself has errors

1186 let SelectModule = class SelectModule {
         ~~~~~~~~~~~~
```

**Comment by petebacondarwin** [comment]
I'm afraid that ngcc is not able to follow symlinks, so this is probably the root of your problem: ngcc is not able to compile `library1` (because it a symlink). So even though the TS compiler can follow the symlink to `library1` when building `library2`, it cannot compile successfully because it `library1` has not been processed by ngcc.

**Comment by yamaha252** [comment]
@petebacondarwin thank you! I understood, you're right. If I use installed `library2` the compiler prepares `library1` because it's in the common node_modules folder. If I use linked one, the compiler tries to use `library1` form the symlink's node_modules folder and there is not prepared by ngcc `library1` package.

**Comment by yamaha252** [comment]
Finally, I solve my problem. I have to have Ivy enabled for development to use `npm/yarn link` as it's by default in the latest angular cli setup. :)

**Comment by mtgibbs** [comment]
> > Just fyi: I have the problems only in conjunction with yarn link or npm link. Oh, and you have to use ngcc after linking.
> 
> @dhubens - I am afraid that ngcc does not process packages that have been symlinked into the `node_modules` directory. This is probably the cause of your errors. Instead generate a packed version of your library and then install that into your application project.

This was my problem when trying to work locally.  When I `npm pack` my library and install the tar via `file:/`, ngcc has no problem and I can proceed.  It's a little annoying because I have to keep a terminal open to keep packing the tar, but its no show-stopper.  Thanks @petebacondarwin

**Comment by yazan-mehrez** [comment]
> As it turns out, for us the issue was caused by the presence of `node_modules` folder inside the `src` folder. Removing this folder made this error go away.
> Hope it helps...

@clement911 Please can you explain which folder exactly you removed?

**Comment by clement911** [comment]
@yazan-mehrez `src/node_modules`. See https://github.com/angular/angular/issues/36060

**Comment by yazan-mehrez** [comment]
> @yazan-mehrez `src/node_modules`. See #36060

@clement911 I don't have this folder, I did all the solutions here but I'm still stuck. flex layout is not working with ivy properly.

**Comment by flash-me** [comment]
@petebacondarwin 

I'm using npm link as described [here](https://github.com/angular/angular/issues/35399#issuecomment-600146117) which creates a link within node_modules and ngcc is able to handle it. **So it works**

cheers
flash :zap:

**Comment by petebacondarwin** [comment]
@flash-me - I am glad that it is working for you. But just to be clear, when ngcc finds a symlink it does not follow it. Now it may be that the CLI is following symlinks and then passing the source of the symlink to ngcc, which may well work.

**Comment by flash-me** [comment]
@petebacondarwin  Good to know that the symlink support is related to cli and not ngcc. I just wanted to mention it. sorry if my formulation sound kinda offensive

cheers 
flash :zap:

**Comment by petebacondarwin** [comment]
No worries @flash-me !

**Comment by JoelKThomas** [comment]
> Not sure this will provide anyone relief or assist with figuring out what the root cause is, but cleaning my cache seemed to help resolve the issue.
> `npm cache clean --force`
> 
> The only other step that I did was I ran a --prod serve after the clean.
> `ng serve --prod`
> 
> I do not know if the clean or --prod thing had anything to do with fixing it for me, or if maybe the combination of them did anything, but I can no longer repro this bug in my project. Note, I can run without --prod and still have successful builds going forward.
> 
> Other than that, I saw this first when adding HttpClientModule as I was going through the tour of heroes tutorial.

This worked for me also

**Comment by HanaaGhasoub** [comment]
Hi,
What is the final solution for that error?
FYI, its working fine when build locally, **but when run build prod it fails**.
Please help me in that as soon as.
![image](https://user-images.githubusercontent.com/30862539/79797298-4cacae00-8357-11ea-9787-82099759868a.png)

My tsconfig.ts
![image](https://user-images.githubusercontent.com/30862539/79797470-939aa380-8357-11ea-843d-2d0c85ddf896.png)

Package.json build scripts
![image](https://user-images.githubusercontent.com/30862539/79797535-b4fb8f80-8357-11ea-827f-489499308cc7.png)
